source/clang/retarget_syscalls.c

```diff
@@ -20,182 +20,91 @@
 
 #include "RTE_Components.h"
 
-#ifdef RTE_CMSIS_Compiler_IO_STDOUT_EVR
-#include "EventRecorder.h"
+#ifdef RTE_CMSIS_Compiler_STDERR
+#include "retarget_stderr.h"
 #endif
 
-#if (defined(RTE_CMSIS_Compiler_IO_STDIN_ITM)  || \
-     defined(RTE_CMSIS_Compiler_IO_STDOUT_ITM) || \
-     defined(RTE_CMSIS_Compiler_IO_STDERR_ITM))
+#ifdef RTE_CMSIS_Compiler_STDIN
+#include "retarget_stdin.h"
+#endif
+
+#ifdef RTE_CMSIS_Compiler_STDOUT
+#include "retarget_stdout.h"
+#endif
 
-/* ITM registers */
-#define ITM_PORT0_U8          (*((volatile uint8_t  *)0xE0000000))
-#define ITM_PORT0_U32         (*((volatile uint32_t *)0xE0000000))
-#define ITM_TER               (*((volatile uint32_t *)0xE0000E00))
-#define ITM_TCR               (*((volatile uint32_t *)0xE0000E80))
+#if (defined(RTE_CMSIS_Compiler_STDIN_ITM)  || \
+     defined(RTE_CMSIS_Compiler_STDOUT_ITM) || \
+     defined(RTE_CMSIS_Compiler_STDERR_ITM))
 
-#define ITM_TCR_ITMENA_Msk    (1UL << 0)
+#ifndef CMSIS_Compiler_CORE_ITM_BUFFER_DISABLE
 
 /*!< Value identifying \ref ITM_RxBuffer is ready for next character. */
-#define ITM_RXBUFFER_EMPTY    0x5AA55AA5
+#define ITM_RXBUFFER_EMPTY  ((int32_t)0x5AA55AA5U)
 
-/*!< Variable to receive characters. */
+/*!< Variable to receive ITM characters. */
 extern
 volatile int32_t ITM_RxBuffer;
 volatile int32_t ITM_RxBuffer = ITM_RXBUFFER_EMPTY;
+#endif /* CMSIS_Compiler_CORE_ITM_BUFFER_DISABLE */
 
-/** \brief  ITM Send Character
-
-    The function transmits a character via the ITM channel 0, and
-    \li Just returns when no debugger is connected that has booked the output.
-    \li Is blocking when a debugger is connected, but the previous character
-        sent has not been transmitted.
-
-    \param [in]     ch  Character to transmit.
-
-    \returns            Character to transmit.
- */
-int32_t ITM_SendChar (int32_t ch);
-int32_t ITM_SendChar (int32_t ch) {
-  if ((ITM_TCR & ITM_TCR_ITMENA_Msk) && /* ITM enabled */
-      (ITM_TER & (1UL << 0)        )) { /* ITM Port #0 enabled */
-    while (ITM_PORT0_U32 == 0);
-    ITM_PORT0_U8 = (uint8_t)ch;
-  }
-  return (ch);
-}
-
-/** \brief  ITM Receive Character
+#endif  /* RTE_CMSIS_Compiler_x_ITM */
```

Comment:
This doesn't look compiler specific to me. Why do we need this here?

Reply:
ITM_RxBuffer? ITM_SendChar and ITM_GetChar are defined in CMSIS but ITM_RxBuffer is external and must be defined somewhere else. So, to enable ITM retargeting, ITM_RxBuffer must exist somewhere. Core components looks like a good candidate, because, where else could it be?

Reply:
As this is not Compiler but channel specific, can it go next to ITM implementation?

Reply:
Yes, it will be moved to ITM implementation.